File: refactor-python/Models/Sala.py
import datetime


class Sala:
    def __init__(self, idSala):
        self.idSala = idSala
        self.reservas = []

    def reservar(self, novaReserva):
        for reservaExistente in self.reservas:
            if novaReserva.isConflito(reservaExistente):
                raise ValueError("Conflito de reserva detectado")
        self.reservas.append(novaReserva)
        return True

    def cancelar(self, idReserva):
        for reserva in self.reservas:
            if reserva.getIdReserva() == idReserva:
                self.reservas.remove(reserva)
                return True
        raise ValueError("Reserva não encontrada")

    def verReservasFuturas(self):
        return [r for r in self.reservas if r.getDataHoraInico() > datetime.now()]

    def getIdSala(self):
        return self.idSala
    def getReservas(self):
        return self.reservas

    def setIdSala(self, idSala):
        self.idSala = idSala
    def gerarIdReserva(self, sistemaDeReservas):
        return f"RES-{len(self.getReservas()) + 1}{sistemaDeReservas.getNumeroSalas() + 1}"

File: src/Models/Sala.java
package Models;

import Interfaces.Reservavel;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class Sala implements Reservavel {
    private String id;
    private List<Reserva> reservas;

    public Sala(String id) {
        this.id = id;
        this.reservas = new ArrayList<>();
    }
    public boolean reservar(Reserva r){
        for(Reserva rr : reservas){
            if(r.isConflito(rr)){
               throw new IllegalArgumentException("Conflito de reserva: " + r.getIdReserva() + " com " + rr.getIdReserva());
            }
        }
        reservas.add(r);
        return true; // Reserva bem-sucedida
    }
    public boolean cancelar(String idReserva) {
        boolean removido = reservas.removeIf(reserva -> reserva.getIdReserva().equals(idReserva));
        if (!removido) {
            System.out.println("Reserva com ID " + idReserva + " não encontrada.");
        }
        return removido;
    }

    public List<Reserva> verReservasFuturas(){
        return reservas.stream()
                .filter(r -> r.getDataHoraInicio().isAfter(java.time.LocalDateTime.now()))
                .collect(Collectors.toList());
    }

    public String getIdSala() {
        return id;
    }
    public void setIdSala(String id) {
        this.id = id;
    }

    public List<Reserva> getReservas(){
        return new ArrayList<>(reservas);
    }
    public String gerarIdReserva(SistemaDeReservas sistemaDeReservas) {
        return "RES-" + reservas.size() + sistemaDeReservas.getNumeroSalas() + 1;
    }
}
